lib/Support/Passes.cpp

```diff
@@ -0,0 +1,44 @@
+//===- Passes.cpp - Helpers for pipeline instrumentation --------*- C++ -*-===//
+//
+// Proprietary and Confidential Software of SiFive Inc. All Rights Reserved.
+// See the LICENSE file for license information.
+// SPDX-License-Identifier: UNLICENSED
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/Passes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "llvm/Support/Format.h"
+
+using namespace mlir;
+using namespace circt;
+
+void VerbosePassInstrumentation::runBeforePass(Pass *pass, Operation *op) {
+  // This assumes that it is safe to log messages to stderr if the operation
+  // is circuit or module op.
+  if (isa<firrtl::CircuitOp, mlir::ModuleOp>(op)) {
```

Comment:
I don't block the PR but ideally we don't want to hardcode this dialect specific condition here. Probably we can define DialectInterface to encode the condition.

Reply:
We can look into this later.

Reply:
This is pretty interesting. I wonder why this is there in the first place 🤔 Is it because of the canonicalizer running on ops in a very fine-grained fashion, where this would print tons of entries to the log?

Reply:
Without the condition `runBeforePass/runAfterPass` would dump logs per FModule/HWModule pass executions. These passes are mostly executed many times in multhreads. Since we are dumping logs to stderr directly, the log breaks down.